dbus_next/service.py

```diff
@@ -431,23 +433,42 @@ def _get_signals(interface):
 
     @staticmethod
     def _fn_result_to_body(result, signature_tree):
+        # TODO: find better way of passing results here instead of list
         out_len = len(signature_tree.types)
         if result is None:
-            body = []
-        elif out_len == 0:
-            raise SignatureBodyMismatchError('Function was not expected to return an argument')
-        elif out_len == 1:
-            body = [result]
-        elif type(result) is not list:
-            raise SignatureBodyMismatchError('Expected function to return a list of arguments')
+            result = []
         else:
-            body = result
+            if out_len == 1:
+                result = [result]
+            else:
+                if type(result) is not list:
+                    raise SignatureBodyMismatchError('Expected signal to return a list of arguments')
 
-        return body
+        if out_len != len(result):
+            raise SignatureBodyMismatchError(
+                "Signature and function return mismatch, expected %s arguments but got %s",
+                (len(signature_tree.types), len(result))
+            )
+
+        body = list(result)
+        fds = []
+
+        def _replace(obj):
+            if getattr(obj, "fileno", False):
```

Comment:
There's `hasattr` but better might be `type(obj) is socket`.

Reply:
I changed this to use hasattr. If I'm not mistaken this doesn't have to be a socket but could for example be an open file.